lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -31,8 +32,33 @@
 from ansible.module_utils.six import string_types
 from ansible.utils.version import SemanticVersion
 
+from collections.abc import Set
 from resolvelib import AbstractProvider
 
+PINNED_CANDIDATE_ATTRS = ('fqcn', 'ver', 'src', 'type')
+
+
+class PinnedCandidateRequests(Set):
```

Comment:
The more I stare at this, the less I understand why an extra set implementation is even necessary. Is it just to filter out non-`_ComputedReqKindsMixin` instances? Why don't we just reject non-conforming objects, in the init? Why don't we use type annotations to protect from this early?
Can we just implement `_ComputedReqKindsMixin.__equals__()` maybe?

If there's a good reason to have it, maybe a docstring and a code comment could be injected.

Reply:
The purpose of this class is to exclude the `signatures` attr when checking if a collection match found on a galaxy server is in the pinned candidates.
`if candidate in self._pinned_candidate_requests:` in `_is_user_requested()` will otherwise always be false if there are galaxy signatures, because the potential Candidate matches have signatures (sourced from the Galaxy server) but the pinned Candidate objects were created from Requirement objects and don't have signatures. The pinned candidates are used to determine if a candidate has been requested by the user so the signatures don't matter anyway in this case.

> Can we just implement `_ComputedReqKindsMixin.__equals__()` maybe?

That's a possible alternative, although it seems really general for this niche case... You don't think we might need to compare all attributes in other situations? And I guess since you're suggesting implementing it in _ComputedReqKindsMixin`, a requirement could equal candidate? It seems like it might break stuff or at least be something to unwittingly trip over.

I can add a doc string.

Reply:
I added the docstring:
`"""Custom set class to store Candidate objects. Excludes the 'signatures' attribute when determining if a Candidate instance is in the set."""`

And an inline comment in `__contains__`:
```
            # Compare Candidate attributes excluding "signatures" since it is
            # unrelated to whether or not a matching Candidate is user-requested.
            # Candidate objects in the set are not expected to have signatures.
```